### PR TITLE
docs: mark matching API and related features as alpha

### PR DESCRIPTION
**Pull request by snowp**
Commit Message: Marks the matching API more clearly as experimental and updates the composite filter security posture to reflect this.
Additional Description:
Risk Level: Low, docs only
Testing: n/a
Docs Changes: n/a 
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: api/envoy/config/common/matcher/v3/matcher.proto
Comment (htuch) [inline api/envoy/config/common/matcher/v3/matcher.proto:25]: Can you add a `[#experimental: ` tag and render this as a warning box? @phlax could you help @snowp with this if possible? It should be doable in https://github.com/envoyproxy/envoy/blob/main/tools/protodoc/protodoc.py.

Also, can https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/security/threat_model#threat-model be updated to explicitly state anything flagged as experimental is out-of-scope?

Comment (phlax) [reply api/envoy/config/common/matcher/v3/matcher.proto:25]: apologies - i misread comment - i can help adding the annotation + parser if its helpful to

File: docs/root/intro/arch_overview/security/threat_model.rst
Comment (htuch) [inline docs/root/intro/arch_overview/security/threat_model.rst:81]: "marked as experimental; experimental features are only supported in trusted deployments and do not qualify for treatment under the threat model below" or something like that.

File: tools/protodoc/protodoc.py
Comment (htuch) [inline tools/protodoc/protodoc.py:166]: "not covered by the threat model" (and link back to the RST for this)

**Review by htuch — COMMENTED** [review]
Just a general comment, we now have multiple notions of WiP:
- This experimental protodoc tag.
- The not-implemented-hide protodoc tag (sort of overlaps but not exactly).
- `alpha` status on the extension.
- `work_in_progress` proto annotation.

Is it clear which to use and where?

**Comment by snowp** [comment]
I think the need for experimental comes in here for core protos that are experimental: the protos live in a file with existing stable protos (maybe it shouldn't have, that way we could have just marked this as alpha/work-in-progress). I also think the WIP tag only works for extension docs, which isn't going to be applicable to these core protos.

Perhaps we should add docs somewhere or attempt to reconcile these variations? For example, we could have a message level WIP annotation instead of this new experimental tag?

**Comment by phlax** [comment]
> Perhaps we should add docs somewhere or attempt to reconcile these variations?

seems like a good idea - but it would be good to get *something* one way or another, so im +1 to adding it for now unless we can resolve upstream quickly

> For example, we could have a message level WIP annotation instead of this new experimental tag?

my feeling is that experimental needs to be distinct from WIP

**Comment by htuch** [comment]
Maybe replace WIP annotations with experimental at file level? Docs coverage in `api/STYLE.md` would be fantastic.

**Review by htuch — COMMENTED** [review]
LGTM modulo `api/STYLE.md` update on the range of experimental/WiP/alpha etc.

File: api/STYLE.md
Comment (alyssawilk) [inline api/STYLE.md:37]: I thought that generally alpha things weren't considered to be in the threat model.  Maybe we should just tag this alpha instead?

@yanavlasov for thoughts

Comment (htuch) [reply api/STYLE.md:37]: This gets to my comment above https://github.com/envoyproxy/envoy/pull/16210#pullrequestreview-649949859. I think the issue is we can only tag as alpha at file level. @snowp points out that sometimes we want to tag at feature level.

Comment (alyssawilk) [reply api/STYLE.md:37]: Right, but can we call them both alpha, rather than alpha and experimental?  Unless there are actual differences, but I think the description above could apply to either so I'd lean towards calling it alpha and shipping :-)

Comment (htuch) [reply api/STYLE.md:37]: Yes, I'm good with more consistent terminology here, that helps reduce some of the confusion.

**Review by alyssawilk — COMMENTED** [review]
oh and PR name/description experimental -> alpha?

File: api/STYLE.md
Comment (snowp) [reply api/STYLE.md:38]: Hmm we still have extension level alpha tagging which differs from this alpha tag which is the only way to mark messages within core as alpha. What kind of change were you thinking about?

File: docs/root/intro/arch_overview/advanced/matching/matching_api.rst
Comment (htuch) [inline docs/root/intro/arch_overview/advanced/matching/matching_api.rst:8]: Nit: s/experimental/alpha/

File: tools/protodoc/protodoc.py
Comment (htuch) [inline tools/protodoc/protodoc.py:166]: FWIW, we have a broader issue which we can resolve after we land this PR. We have both implementation and API alpha/experimental/unknown robustness status. An API is not really insecure in and of itself, this is an implementation attribute.

We really need a way to mark in the protos client specific implementation status details, since Envoy vs. gRPC support for a feature might have different implementation status.

I started to explore this in https://github.com/envoyproxy/envoy/issues/11085, but it's not really something I have b/w for. Maybe @phlax this can be part of the docs/tooling roadmap?

CC  @markdroth

**Review by htuch — COMMENTED** [review]
LGTM modulo nit. Will leave final sign-off/merge to @alyssawilk

File: api/STYLE.md
Comment (alyssawilk) [reply api/STYLE.md:38]: I was trying to resolve the two, and have alpha tagging for extensions be functionally the same as alpha tagging for messages and fields, which is basically that both are WIP, not stable APIs and less trusted code.

File: tools/protodoc/protodoc.py
Comment (markdroth) [reply tools/protodoc/protodoc.py:166]: I agree, it would be nice to have some generic solution for this kind of thing.  Matt has been saying for a while that we need a way to have client-specific docs in the proto files, not just for API status but for any type of documentation.  Although as the number of xDS clients grows, I'm not sure how feasible it will be to document all of the in the same proto files.  But this is definitely something we need to come up with a good strategy for.

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**
